### PR TITLE
[Cosmos] Reconfigure retry policy

### PR DESCRIPTION
**Pull request by annatisch**
Adding support for connection retry as a Pipeline policy rather than adding a urllib Retry object to the session.
Should be 99% backwards compatible.
Tests will require this fix for azure-core in order to pass: PR #7542

### CODE REVIEW

**Comment by adxsdk6** [comment]
Can one of the admins verify this patch?

File: sdk/cosmos/azure-cosmos/azure/cosmos/_cosmos_client_connection.py [hunk 1]
```diff
@@ -151,15 +153,22 @@ def __init__(
...
+            # Convert a urllib3 retry policy to a Pipeline policy
+            retry_policy = RetryPolicy(
+                retry_total=self.connection_policy.ConnectionRetryConfiguration.total,
+                retry_connect=self.connection_policy.ConnectionRetryConfiguration.connect,
+                retry_read=self.connection_policy.ConnectionRetryConfiguration.read,
+                retry_status=self.connection_policy.ConnectionRetryConfiguration.status,
+                retry_backoff_max=self.connection_policy.ConnectionRetryConfiguration.BACKOFF_MAX,
+                retry_on_status_codes=list(self.connection_policy.ConnectionRetryConfiguration.status_forcelist),
+                retry_backoff_factor=self.connection_policy.ConnectionRetryConfiguration.backoff_factor
+            )
```
Comment (bryevdv) [inline sdk/cosmos/azure-cosmos/azure/cosmos/_cosmos_client_connection.py]: could type check for urllib3 retry and raise a ValueError if it's not

**Comment by bryevdv** [comment]
one small comment, LGTM

File: sdk/cosmos/azure-cosmos/azure/cosmos/errors.py [hunk 2]
```diff
@@ -63,3 +63,13 @@ class CosmosResourceExistsError(ResourceExistsError, CosmosHttpResponseError):
...
+
+
+class ClientTimeoutError(AzureError):
+    """An operation failed to complete within the specified timeout."""
+
+    def __init__(self, **kwargs):
+        message = "Client operation failed to complete within specified timeout."
+        self.response = None
+        self.history = None
+        super(ClientTimeoutError, self).__init__(message, **kwargs)
```
Comment (bryevdv) [inline sdk/cosmos/azure-cosmos/azure/cosmos/errors.py]: Every other class in this module has `Cosmos` in the name. Is this timeout error really specific to cosmos? If so, suggest a similar name for consistency

**Comment by annatisch** [comment]
Ping @johanste, @srinathnarayanan :)

File: sdk/cosmos/azure-cosmos/azure/cosmos/_retry_utility.py [hunk 3]
```diff
@@ -64,6 +67,8 @@ def Execute(client, global_endpoint_manager, function, *args, **kwargs):
     )
     while True:
         try:
+            client_timeout = kwargs.get('timeout')
+            start_time = time.time()
```
Comment (johanste) [inline sdk/cosmos/azure-cosmos/azure/cosmos/_retry_utility.py]: Should this be outside the loop?

Comment (annatisch) [reply sdk/cosmos/azure-cosmos/azure/cosmos/_retry_utility.py]: Thanks @johanste 
I don't think so - we calculate the time spent for each loop iteration and deduct it from the absolute timeout, which we then feed back into kwargs on line 122 - so I think it's correct to do this with each iteration

**Pull request merged**